### PR TITLE
fix: GTID Truncation Issue and Improve Consistency in Checkpoint Management

### PR DESCRIPTION
**Pull request by yyuuttaaoo**
This patch addresses an issue where the method used to retrieve GTIDs (SHOW GLOBAL VARIABLES LIKE 'gtid_executed') causes truncation to 1024 bytes, leading to invalid GTIDs. Consequently, the input_canal would fall back to the File Position sync method. However, if the old checkpoint contained a GTID, it was not cleared, resulting in the possibility of using the wrong GTID upon plugin reload. This could cause data loss as the plugin would always start from the latest File Position.

To fix this, the patch implements the following changes:

Use SELECT @@global.gtid_executed to retrieve the GTID, thereby preventing truncation issues.
Ensure that the GTID in the checkpoint is removed if the File Position sync method is in use.
Upgrade the canal library from version 1.3 to 1.8 to incorporate the latest features and improvements.
These updates enhance the reliability and consistency of GTID management and prevent potential data loss upon plugin reload.

### CODE REVIEW

File: plugins/input/canal/input_canal.go [hunk 1]
```diff
@@ -762,13 +775,11 @@ ForBlock:
 		case <-sc.shutdown:
 			logger.Info(sc.context.GetRuntimeContext(), "service_canal quit because of shutting down, checkpoint",
 				sc.checkpoint)
-			sc.canal.Close()
+			sc.canal.Close() // will trigger OnPosSynced with force=true
 			<-sc.lastErrorChan
-			sc.saveCheckPoint()
```
Comment (messixukejia) [inline plugins/input/canal/input_canal.go:767]: shutdow不需要save？

File: plugins/input/canal/input_canal.go [hunk 2]
```diff
@@ -350,13 +344,14 @@ func (sc *ServiceCanal) columnValueToString(column *schema.TableColumn, rowVal i
...
-	// logger.Debug("OnRow", *e, "GTID", sc.nextRowEventGTID)
+	logger.Debug(sc.context.GetRuntimeContext(), "[CANAL_DEBUG] host", sc.Host, "db", e.Table.Schema, "table", e.Table.Name, "action", e.Action, "GTID", sc.nextRowEventGTID)
 	sc.rowCounter.Add(1)
 	values := map[string]string{"_host_": sc.Host, "_db_": e.Table.Schema, "_table_": e.Table.Name, "_event_": "row_" + e.Action, "_id_": strconv.Itoa(sc.checkpoint.ID)}
+	nextOffsetString := strconv.FormatUint(uint64(e.Header.LogPos), 10)
 	if sc.EnableGTID {
 		values["_gtid_"] = sc.nextRowEventGTID
 		values["_filename_"] = sc.checkpoint.FileName
-		values["_offset_"] = sc.lastOffsetString
+		values["_offset_"] = nextOffsetString
```
Comment (messixukejia) [inline plugins/input/canal/input_canal.go:354]: 测试过数据连续行吗？特别是logtail退出场景

**Review by messixukejia — CHANGES_REQUESTED** [review]

**Comment by yyuuttaaoo** [comment]
# 简要测试报告

## 测试获取超长GTID
使用测试RDS实例，不断进行可用区切换创建新的GTID

image

长度超过1024后，停止ilogtail并开始写入数据，然后启动ilogtail采集。预期应该使用checkpoint中的进度，将停止期间的数据均采集上来。

使用老版本将无法正确获取GTID，报错信息如下
```
2024-07-27 11:09:54 [WRN] [input_canal.go:747] [Start] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]      AlarmType:CANAL_START_ALARM     Call getLatestGTID failed, error:ParseGTIDSet(0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,
0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,
1280318a-4a62-11ef-9227-00163e042bc7:1-55,
20370918-4b14-11ef-a4c9-00163e041923:1-1641,
30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,
344ec336-4996-11ef-b535-00163e01f5fb:1-98,
3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,
49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,
56556036-4a69-11ef-8779-00163e041923:1-1593,
59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,
758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,
92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,
9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,
9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,
9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,
caad3ef8-4b64-11ef-9131-00163e041923:1-2663,
caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,
cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,
ce68989a-4a50-11ef-875a-00163e041923:1-61,
d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,
d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,
d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,
d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,
f0eac814-4a52-11) failed: invalid GTID format, must UUID:interval[:interval]    logstore:mysqlbinlog    config:##1.0##xunfei-test-wlcb$canal_plugin_config
```
新版本可以正常获取GTID，信息如下
```
2024-07-27 11:20:13 [INF] [input_canal.go:767] [Start] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]      Get latest checkpoint GTID: 0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,1280318a-4a62-11ef-9227-00163e042bc7:1-55,20370918-4b14-11ef-a4c9-00163e041923:1-1641,30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,344ec336-4996-11ef-b535-00163e01f5fb:1-98,3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,43597e06-4bc5-11ef-9f49-00163e041923:1-40,49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,56556036-4a69-11ef-8779-00163e041923:1-1593,59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,caad3ef8-4b64-11ef-9131-00163e041923:1-2663,caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,ce68989a-4a50-11ef-875a-00163e041923:1-61,d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,f0eac814-4a52-11ef-a21b-00163e042b53:1-271 Position: (, 0)
2024-07-27 11:20:13 [INF] [input_canal.go:566] [runCanalByGTID] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]     start canal from 0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,1280318a-4a62-11ef-9227-00163e042bc7:1-55,20370918-4b14-11ef-a4c9-00163e041923:1-1641,30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,344ec336-4996-11ef-b535-00163e01f5fb:1-98,3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,43597e06-4bc5-11ef-9f49-00163e041923:1-40,49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,56556036-4a69-11ef-8779-00163e041923:1-1593,59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,caad3ef8-4b64-11ef-9131-00163e041923:1-2663,caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,ce68989a-4a50-11ef-875a-00163e041923:1-61,d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,f0eac814-4a52-11ef-a21b-00163e042b53:1-271 with GTID mode
```

## 验证checkpoint正确性
停止ilogtail，checkpoint会在退出时记录
```
2024-07-27 11:21:58 [INF] [plugin_export.go:179] [HoldOn] Hold on:start flag:1
2024-07-27 11:21:58 [DBG] [input_canal.go:500] [OnPosSynced] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]        [CANAL_DEBUG] OnPosSynced: 0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,1280318a-4a62-11ef-9227-00163e042bc7:1-55,20370918-4b14-11ef-a4c9-00163e041923:1-1641,30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,344ec336-4996-11ef-b535-00163e01f5fb:1-98,3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,43597e06-4bc5-11ef-9f49-00163e041923:1-48,49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,56556036-4a69-11ef-8779-00163e041923:1-1593,59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,caad3ef8-4b64-11ef-9131-00163e041923:1-2663,caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,ce68989a-4a50-11ef-875a-00163e041923:1-61,d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,f0eac814-4a52-11ef-a21b-00163e042b53:1-271
2024-07-27 11:21:58 [INF] [input_canal.go:778] [Start] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]      service_canal quit because of shutting down, checkpoint:{0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,1280318a-4a62-11ef-9227-00163e042bc7:1-55,20370918-4b14-11ef-a4c9-00163e041923:1-1641,30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,344ec336-4996-11ef-b535-00163e01f5fb:1-98,3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,43597e06-4bc5-11ef-9f49-00163e041923:1-48,49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,56556036-4a69-11ef-8779-00163e041923:1-1593,59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,caad3ef8-4b64-11ef-9131-00163e041923:1-2663,caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,ce68989a-4a50-11ef-875a-00163e041923:1-61,d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,f0eac814-4a52-11ef-a21b-00163e042b53:1-271 mysql-bin.000057 22593 1}
```
向mysql数据库插入10条数据
启动ilogtail，，checkpoint会在启动时加载
```
2024-07-27 11:22:30 [INF] [input_canal.go:744] [Start] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]      Checkpoint initialized: {0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,1280318a-4a62-11ef-9227-00163e042bc7:1-55,20370918-4b14-11ef-a4c9-00163e041923:1-1641,30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,344ec336-4996-11ef-b535-00163e01f5fb:1-98,3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,43597e06-4bc5-11ef-9f49-00163e041923:1-48,49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,56556036-4a69-11ef-8779-00163e041923:1-1593,59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,caad3ef8-4b64-11ef-9131-00163e041923:1-2663,caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,ce68989a-4a50-11ef-875a-00163e041923:1-61,d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,f0eac814-4a52-11ef-a21b-00163e042b53:1-271 mysql-bin.000057 22593 1}
2024-07-27 11:22:30 [INF] [input_canal.go:566] [runCanalByGTID] [##1.0##xunfei-test-wlcb$canal_plugin_config/1,mysqlbinlog]     start canal from 0e275c3a-4b0c-11ef-8240-00163e042b53:1-225,0e67b598-4a4c-11ef-801f-00163e042bc7:1-68,1280318a-4a62-11ef-9227-00163e042bc7:1-55,20370918-4b14-11ef-a4c9-00163e041923:1-1641,30ff05b2-4a64-11ef-9fb2-00163e012c29:1-148,344ec336-4996-11ef-b535-00163e01f5fb:1-98,3a8a9542-4aa5-11ef-a6fc-00163e012c29:1-1882,43597e06-4bc5-11ef-9f49-00163e041923:1-48,49d7c63e-4b52-11ef-9a4b-00163e041923:1-271,56556036-4a69-11ef-8779-00163e041923:1-1593,59ca159e-4996-11ef-a1d6-00163e042b53:1-4811,758457e8-4bc0-11ef-aa26-00163e04cdf8:1-135,92eb752e-4a5c-11ef-8f0f-00163e041923:1-89,9b0ff65a-4aa0-11ef-9e77-00163e042b53:1-55,9f5ba6be-4ae6-11ef-aeaa-00163e041923:1-1083,9f820ab8-4a4e-11ef-826f-00163e012c29:1-57,caad3ef8-4b64-11ef-9131-00163e041923:1-2663,caf7f514-4b5b-11ef-81f5-00163e042b53:1-251,cbe7c868-4a49-11ef-a5fc-00163e042b53:1-60,ce68989a-4a50-11ef-875a-00163e041923:1-61,d09d5e8e-4a3c-11ef-8af3-00163e041923:1-373,d1d04f5a-4a5f-11ef-bacd-00163e042b53:1-57,d2314182-4aa2-11ef-9e91-00163e042bc7:1-61,d7799cc0-4b4c-11ef-859c-00163e042b53:1-150,f0eac814-4a52-11ef-a21b-00163e042b53:1-271 with GTID mode
```

完美续上，采集10条数据
image

File: plugins/input/canal/input_canal.go (hunk 2 above)
Comment (yyuuttaaoo) [reply plugins/input/canal/input_canal.go:354]: 已补充测试报告

File: plugins/input/canal/input_canal.go (hunk 1 above)
Comment (yyuuttaaoo) [reply plugins/input/canal/input_canal.go:767]: 不需要，如注释// will trigger OnPosSynced with force=true所描述的，退出会在canal框架触发OnPosSync事件

**Review by messixukejia — APPROVED** [review]

**Pull request merged**